File: 337. House Robber III/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode *, int> memory;

    int rob(TreeNode *root) {
        if (root == nullptr) {
            return 0;
        }
        if (memory.find(root) != memory.end()) {
            return memory[root];
        }
        int subSum = 0;
        if (root->left != nullptr) {
            subSum += rob(root->left->left) + rob(root->left->right);
        }
        if (root->right != nullptr) {
            subSum += rob(root->right->left) + rob(root->right->right);
        }
        int result = max(root->val + subSum, rob(root->left) + rob(root->right));
        memory[root] = result;
        return result;
    }
};

File: 337. House Robber III/Python/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def __init__(self):
        self.memory = {}

    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        if root in self.memory:
            return self.memory[root]
        sub_sum = 0
        if root.left:
            sub_sum += self.rob(root.left.left) + self.rob(root.left.right)
        if root.right:
            sub_sum += self.rob(root.right.left) + self.rob(root.right.right)
        result = max(root.val + sub_sum, self.rob(root.left) + self.rob(root.right))
        self.memory[root] = result
        return result
